activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -789,10 +789,14 @@ def initialize_type_map(m) # :nodoc:
         register_integer_type m, %r(^tinyint)i,   limit: 1
 
         m.alias_type %r(tinyint\(1\))i,  'boolean' if emulate_booleans
-        m.alias_type %r(set)i,           'varchar'
         m.alias_type %r(year)i,          'integer'
         m.alias_type %r(bit)i,           'binary'
 
+        m.register_type(%r(set)i) do |sql_type|
+          limit = sql_type[/^set\((.+)\)/i, 1].length
+          MysqlString.new(limit: limit)
+        end
```

Comment:
Why do we need to change this from an alias? If you want a higher precedence just moving the registration should be enough.